app/src/test/java/org/mozilla/fenix/components/metrics/AppStartupTypeTelemetryTest.kt

```diff
@@ -0,0 +1,118 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.metrics
+
+import io.mockk.MockKAnnotations
+import io.mockk.impl.annotations.RelaxedMockK
+import io.mockk.verify
+import org.junit.Before
+import org.junit.Test
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.HOT
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.WARM
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.COLD
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.COLD_SAVED_INSTANCE
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.WARM_SAVED_INSTANCE
+
+class AppStartupTypeTelemetryTest {
+
+    @RelaxedMockK
+    private lateinit var metricController: MetricController
+
+    private lateinit var appStartupTypeTelemetry: AppStartupTypeTelemetry
+
+    @Before
+    fun setup() {
+        MockKAnnotations.init(this)
+        appStartupTypeTelemetry = AppStartupTypeTelemetry(metricController)
+    }
+
+    @Test
+    fun `when FenixApplication and HomeActivity is created without savedInstance, then metric reports COLD startup type`() {
+        appStartupTypeTelemetry.onFenixApplicationOnCreate()
+        appStartupTypeTelemetry.onHomeActivityOnCreate(false)
+
+        val validType = COLD
+
+        verify(exactly = 1) { metricController.track(Event.AppStartupType(validType)) }
+    }
+
+    @Test
+    fun `when FenixApplication and HomeActivity is created with savedInstance, then metric reports COLD_SAVED_INSTANCE startup type`() {
+        appStartupTypeTelemetry.onFenixApplicationOnCreate()
+        appStartupTypeTelemetry.onHomeActivityOnCreate(true)
+
+        val validType = COLD_SAVED_INSTANCE
+
+        verify(exactly = 1) { metricController.track(Event.AppStartupType(validType)) }
+    }
+
+    @Test
+    fun `when application exists, and HomeActivity is created without savedInstanceState, then metric reports WARM startup type`() {
+        appStartupTypeTelemetry.onHomeActivityOnCreate(false)
```

Comment:
This code does not represent `WHEN application exists` – it represents "when this class is created and HomeActivity.onCreate` is called, which is never expected to happen. Let's fix the test to look at application being created (i.e. the name is fine)

When I'm writing test names (and thus designing the test at a high level), I'm careful to make sure my test name represents the code. When I am, I might happen to notice, in this case, 1) that my test isn't testing what I thought it was and needs to change and 2) there's a code path I didn't think of.

wrt 2, ideally we would throw if this unexpected code path occurs but it's probably not worth complicating the code to make that happen (because then we'd need to handle `onRestart`s impossible states too...). fwiw, if we wanted to solve this in a generic way, a state machine is a good way to do it because it doesn't require special-cased code to validate current state. But again, I don't think it's worth the complexity in this case.